src/orion/storage/track.py

```diff
@@ -0,0 +1,464 @@
+        metadata['submit_time'] = to_json(trial.submit_time)
+        metadata['end_time'] = to_json(trial.end_time)
+        metadata['worker'] = trial.worker
+        metadata['metric_types'] = {remove_leading_slash(p.name): p.type for p in trial.results}
+        metadata['metric_types'][self.objective] = 'objective'
+        metadata['heartbeat'] = stamp
+
+        metrics = defaultdict(list)
+        for p in trial.results:
+            metrics[p.name] = p.value
+
+        self.current_trial = self.backend.new_trial(TrackTrial(
+            _hash=trial.hash_name,
+            status=get_track_status('new'),
+            project_id=self.project.uid,
+            group_id=self.group.uid,
+            parameters={p.name: p.value for p in trial.params},
+            metadata=metadata,
+            metrics=metrics
+        ))
+        return TrialAdapter(self.current_trial, objective=self.objective)
+
+    def register_lie(self, trial):
+        """Register a *fake* trial created by the strategist.
+
+        The main difference between fake trial and original ones is the addition of a fake objective
+        result, and status being set to completed. The id of the fake trial is different than the id
+        of the original trial, but the original id can be computed using the hashcode on parameters
+        of the fake trial. See mod:`orion.core.worker.strategy` for more information and the
+        Strategist object and generation of fake trials.
+
+        Parameters
+        ----------
+        trial: `Trial` object
+            Fake trial to register in the database
+
+        """
+        pass
+
+    def fetch_trials(self, query, *args, **kwargs):
+        """Fetch all the trials that match the query"""
+        def sort_key(item):
+            submit_time = item.submit_time
+            if submit_time is None:
+                return 0
+            return submit_time
+
+        query = to_json(query)
+
+        new_query = {}
+        for k, v in query.items():
+            if k == 'experiment':
+                new_query['group_id'] = v
+
+            elif k == 'heartbeat':
+                new_query['metadata.heartbeat'] = v
+
+            elif k == '_id':
+                new_query['uid'] = v
+
+            elif k == 'end_time':
+                new_query['metadata.end_time'] = v
+
+            elif k == 'status' and isinstance(v, str):
+                new_query['status'] = get_track_status(v)
+
+            else:
+                new_query[k] = v
+
+        trials = [
+            TrialAdapter(t, objective=self.objective) for t in self.backend.fetch_trials(new_query)
+        ]
+        trials.sort(key=sort_key)
+        return trials
+
+    _ignore_updates_for = {'results', 'params', '_id'}
+
+    def update_trial(self, trial, where=None, **kwargs):
+        """Update the fields of a given trials
+
+        Parameters
+        ----------
+        trial: Trial
+            Trial object to update
+
+        where: Optional[dict]
```

Comment:
The query constraint is not used.

Reply:
So it actually does not need the constraint because all the information needed to perform the update is stored in `trial` and track cannot do batch update

Reply:
That means it cannot work with Orion. We need the query to ensure we do not reserve trials twice. Isn't there a global lock in Track that we could use for the time of the updates?

Reply:
track adds a version id for each records, they should match else it means the record was modified by something else